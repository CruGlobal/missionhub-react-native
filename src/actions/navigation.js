import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
