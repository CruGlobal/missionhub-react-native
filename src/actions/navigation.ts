/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
