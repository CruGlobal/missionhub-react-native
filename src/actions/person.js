import callApi, { REQUESTS } from './api';
import { PERSON_FIRST_NAME_CHANGED, PERSON_LAST_NAME_CHANGED } from '../constants';


export function personFirstNameChanged(firstName) {
  return {
    type: PERSON_FIRST_NAME_CHANGED,
    personFirstName: firstName,
  };
}

export function personLastNameChanged(lastName) {
  return {
    type: PERSON_LAST_NAME_CHANGED,
    personLastName: lastName,
  };
}

export function saveNotes(personId, notes, firstSave) {
  return (dispatch, getState) => {
    const myId = getState().auth.personId;
    if (!personId || !notes) {
      return Promise.reject('InvalidData', personId, notes);
    }

    const bodyData = {
      data: {
        type: 'person_note',
        attributes: {
          content: notes,
        },
      },
      relationships: {
        person: {
          data: {
            type: 'person',
            id: personId,
          },
        },
      },
      user: {
        data: {
          type: 'user',
          id: myId,
        },
      },
    };

    const query = {};

    if (firstSave) return dispatch(callApi(REQUESTS.ADD_PERSON_NOTES, query, bodyData));
    return dispatch(callApi(REQUESTS.UPDATE_PERSON_NOTES, query, bodyData));
  };
}