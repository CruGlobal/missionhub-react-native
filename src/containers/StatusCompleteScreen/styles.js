import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.primaryColor,
  },
  content: {
    marginTop: 50,
    paddingHorizontal: 35,
  },
  text: {
    fontSize: 24,
    color: theme.white,
    marginBottom: 25,
  },
  button: {
    padding: 13,
    borderColor: theme.secondaryColor,
    borderWidth: 1,
    borderRadius: 25,
    alignSelf: 'stretch',
    marginTop: 15,
  },
  buttonText: {
    fontSize: 14,
  },
});
