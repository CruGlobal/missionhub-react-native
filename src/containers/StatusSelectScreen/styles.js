import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.white,
  },
  headerButton: {
    marginHorizontal: 10,
    marginTop: 5,
  },
  headerButtonText: {
    fontSize: 14,
    letterSpacing: 2,
  },
  row: {
    paddingHorizontal: 15,
  },
  text: {
    fontSize: 16,
  },
  selected: {
    color: theme.secondaryColor,
  },
  icon: {
    color: theme.secondaryColor,
  },
});
