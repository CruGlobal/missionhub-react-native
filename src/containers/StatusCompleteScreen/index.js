import React, { Component } from 'react';
import { View } from 'react-native';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

import { Flex, Text, Button } from '../../components/common';
import Header from '../Header';
import BackButton from '../BackButton';

import styles from './styles';

@translate('statusComplete')
class StatusCompleteScreen extends Component {
  cancel = () => {
    return false;
  };

  complete = () => {
    return false;
  };

  render() {
    const { t, me, person } = this.props;

    return (
      <View style={styles.container}>
        <Header left={<BackButton />} shadow={false} />

        <Flex value={1} align="stretch" style={styles.content}>
          <Text style={styles.text}>
            {t('continue', {
              userName: me.first_name,
              statusName: person.first_name,
            })}
          </Text>
          <Button
            type="transparent"
            onPress={this.complete}
            text={t('totally').toUpperCase()}
            style={styles.button}
            buttonTextStyle={styles.buttonText}
          />
          <Button
            type="transparent"
            onPress={this.cancel}
            text={t('nope').toUpperCase()}
            style={styles.button}
            buttonTextStyle={styles.buttonText}
          />
        </Flex>
      </View>
    );
  }
}

StatusCompleteScreen.propTypes = {
  person: PropTypes.object.isRequired,
};

const mapStateToProps = ({ auth }, { navigation }) => ({
  ...(navigation.state.params || {}),
  me: auth.person,
});

export default connect(mapStateToProps)(StatusCompleteScreen);
export const STATUS_COMPLETE_SCREEN = 'nav/STATUS_COMPLETE';
