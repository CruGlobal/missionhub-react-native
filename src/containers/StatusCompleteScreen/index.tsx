import React, { Component } from 'react';
import { View } from 'react-native';
import { connect } from 'react-redux-legacy';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';

import { Flex, Text, Button } from '../../components/common';
import { navigatePush, navigateBack } from '../../actions/navigation';
import Header from '../../components/Header';
import BackButton from '../BackButton';
import { STATUS_REASON_SCREEN } from '../StatusReasonScreen';

import styles from './styles';

// @ts-ignore
@withTranslation('statusComplete')
class StatusCompleteScreen extends Component {
  onSubmitReason = () => {
    // @ts-ignore
    this.props.dispatch(navigateBack(4));
  };

  cancel = () => {
    // @ts-ignore
    const { dispatch, person, contactAssignment, organization } = this.props;
    dispatch(
      navigatePush(STATUS_REASON_SCREEN, {
        person,
        organization,
        contactAssignment,
        onSubmit: this.onSubmitReason,
      }),
    );
  };

  complete = () => {
    // @ts-ignore
    const { dispatch } = this.props;
    dispatch(navigateBack(2));
  };

  render() {
    // @ts-ignore
    const { t, me, person } = this.props;

    return (
      <View style={styles.container}>
        <Header left={<BackButton />} />
        <Flex value={1} align="stretch" style={styles.content}>
          <Text style={styles.text}>
            {t('continue', {
              userName: me.first_name,
              statusName: person.first_name,
            })}
          </Text>
          <Button
            type="transparent"
            onPress={this.complete}
            text={t('totally').toUpperCase()}
            style={styles.button}
            buttonTextStyle={styles.buttonText}
          />
          <Button
            type="transparent"
            onPress={this.cancel}
            text={t('nope').toUpperCase()}
            style={styles.button}
            buttonTextStyle={styles.buttonText}
          />
        </Flex>
      </View>
    );
  }
}

// @ts-ignore
StatusCompleteScreen.propTypes = {
  person: PropTypes.object.isRequired,
  organization: PropTypes.object.isRequired,
};

// @ts-ignore
const mapStateToProps = ({ auth }, { navigation }) => ({
  ...(navigation.state.params || {}),
  me: auth.person,
});

export default connect(mapStateToProps)(StatusCompleteScreen);
export const STATUS_COMPLETE_SCREEN = 'nav/STATUS_COMPLETE';
