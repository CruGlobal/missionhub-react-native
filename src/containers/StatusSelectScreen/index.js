import React, { Component } from 'react';
import { View } from 'react-native';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

import { Flex, Text, Button, Touchable, Icon } from '../../components/common';
import Header from '../Header';
import {
  contactAssignmentSelector,
  orgPermissionSelector,
  personSelector,
} from '../../selectors/people';
import { updateFollowupStatus } from '../../actions/person';
import { navigatePush, navigateBack } from '../../actions/navigation';
import { STATUS_COMPLETE_SCREEN } from '../StatusCompleteScreen';

import styles from './styles';

@translate('statusSelect')
class StatusSelectScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: props.status,
    };
  }

  select = status => {
    this.setState({ selected: status });
  };

  submit = async () => {
    const {
      dispatch,
      person,
      organization,
      contactAssignment,
      orgPermission,
      status,
    } = this.props;
    const { selected } = this.state;
    if (status === selected) {
      dispatch(navigateBack());
      return;
    }
    await dispatch(updateFollowupStatus(person, orgPermission.id, selected));
    if (selected === 'completed' || selected === 'do_not_contact') {
      dispatch(
        navigatePush(STATUS_COMPLETE_SCREEN, {
          organization,
          person,
          contactAssignment,
        }),
      );
    }
  };

  renderItem(type) {
    const { t } = this.props;
    const { selected } = this.state;
    return (
      <Touchable pressProps={[type]} onPress={this.select} style={styles.row}>
        <Text style={[styles.text, selected === type ? styles.selected : null]}>
          {t(type)}
        </Text>
        {selected === type ? (
          <Icon
            name="checkIcon"
            size={16}
            type="MissionHub"
            style={styles.icon}
          />
        ) : null}
      </Touchable>
    );
  }

  navigateBack = () => this.props.dispatch(navigateBack());

  render() {
    const { t } = this.props;

    return (
      <View style={styles.container}>
        <Header
          left={
            <Button
              text={t('cancel').toUpperCase()}
              type="transparent"
              onPress={this.navigateBack}
              style={styles.headerButton}
              buttonTextStyle={[
                styles.headerButtonText,
                styles.headerButtonTextCancel,
              ]}
            />
          }
          right={
            <Button
              text={t('done').toUpperCase()}
              type="transparent"
              onPress={this.submit}
              style={styles.headerButton}
              buttonTextStyle={[
                styles.headerButtonText,
                styles.headerButtonTextDone,
              ]}
            />
          }
          title={t('header')}
        />

        <Flex value={1} align="stretch">
          {this.renderItem('uncontacted')}
          {this.renderItem('attempted_contact')}
          {this.renderItem('contacted')}
          {this.renderItem('completed')}
          {this.renderItem('do_not_contact')}
        </Flex>
      </View>
    );
  }
}

StatusSelectScreen.propTypes = {
  person: PropTypes.object.isRequired,
  organization: PropTypes.object.isRequired,
};

export const mapStateToProps = ({ auth, people }, { navigation }) => {
  const navParams = navigation.state.params || {};
  const orgId = navParams.organization && navParams.organization.id;
  const person =
    personSelector({ people }, { personId: navParams.person.id, orgId }) ||
    navParams.person;
  const organization = navParams.organization;
  const orgPermission = orgPermissionSelector(null, {
    person,
    organization,
  });

  return {
    ...navParams,
    person,
    organization,
    orgPermission: orgPermission,
    contactAssignment: contactAssignmentSelector({ auth }, { person, orgId }),
    status: orgPermission.followup_status,
  };
};

export default connect(mapStateToProps)(StatusSelectScreen);
export const STATUS_SELECT_SCREEN = 'nav/STATUS_SELECT';
