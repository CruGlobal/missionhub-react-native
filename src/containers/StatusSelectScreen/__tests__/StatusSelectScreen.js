/* eslint max-lines-per-function: 0 */

import React from 'react';

import StatusSelectScreen, { mapStateToProps } from '..';

import {
  createMockStore,
  renderShallow,
  testSnapshotShallow,
  createMockNavState,
} from '../../../../testUtils';
import { updateFollowupStatus } from '../../../actions/person';
import * as navigation from '../../../actions/navigation';
import { STATUS_COMPLETE_SCREEN } from '../../StatusCompleteScreen';
import {
  contactAssignmentSelector,
  orgPermissionSelector,
  personSelector,
} from '../../../selectors/people';

jest.mock('../../../actions/person', () => ({
  updateFollowupStatus: jest.fn(() => Promise.resolve()),
}));
jest.mock('../../../selectors/people');

const store = createMockStore({});

const followupStatus = 'uncontacted';
const orgPermission = {
  id: 'orgPerm1',
  organization_id: '1',
  followup_status: followupStatus,
};
const contactAssignment = { id: 'assignment1' };

const person = {
  id: 'person1',
  full_name: 'Person One',
  organizational_permissions: [orgPermission],
};
const organization = { id: '1', name: 'Test Org' };

contactAssignmentSelector.mockReturnValue(contactAssignment);
orgPermissionSelector.mockReturnValue(orgPermission);
personSelector.mockReturnValue(person);

describe('mapStateToProps', () => {
  it('provides props correctly', () => {
    expect(
      mapStateToProps(
        {
          auth: {
            person: {
              id: '1',
            },
          },
          people: {
            allByOrg: {
              [organization.id]: person,
            },
          },
        },
        {
          navigation: {
            state: {
              params: {
                organization,
                person,
              },
            },
          },
        },
      ),
    ).toEqual({
      person,
      organization,
      orgPermission,
      contactAssignment,
      status: followupStatus,
    });
  });
});

describe('StatusSelectScreen', () => {
  beforeEach(() => {
    navigation.navigatePush.mockClear();
  });
  let instance;

  const component = (
    <StatusSelectScreen
      navigation={createMockNavState({
        person,
        organization,
      })}
    />
  );
  navigation.navigateBack = jest.fn();
  navigation.navigatePush = jest.fn();

  const testSubmit = async type => {
    instance = renderShallow(component, store).instance();
    instance.setState({ selected: type });
    await instance.submit();
    expect(updateFollowupStatus).toHaveBeenCalledWith(
      person,
      orgPermission.id,
      type,
    );
  };

  it('should render correctly', () => {
    testSnapshotShallow(component, store);
  });

  it('should navigate back', async () => {
    const instance = renderShallow(component, store).instance();
    await instance.submit();
    expect(navigation.navigateBack).toHaveBeenCalled();
  });

  it('should update status to attempted_contact', async () => {
    await testSubmit('attempted_contact');
  });

  it('should update status to completed', async () => {
    await testSubmit('completed');

    expect(navigation.navigatePush).toHaveBeenCalledWith(
      STATUS_COMPLETE_SCREEN,
      {
        organization,
        person,
        contactAssignment,
      },
    );
  });

  it('should update status to do_not_contact', async () => {
    await testSubmit('do_not_contact');

    expect(navigation.navigatePush).toHaveBeenCalledWith(
      STATUS_COMPLETE_SCREEN,
      {
        organization,
        person,
        contactAssignment,
      },
    );
  });

  it('set the state to contacted', () => {
    const instance = renderShallow(component, store).instance();
    const status = 'contacted';
    instance.select(status);
    expect(instance.state.selected).toEqual(status);
  });
});
