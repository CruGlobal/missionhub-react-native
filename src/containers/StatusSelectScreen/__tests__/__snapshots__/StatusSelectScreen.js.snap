// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusSelectScreen should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#007398",
      "flex": 1,
    }
  }
>
  <Header
    left={
      <Button
        buttonTextStyle={
          Array [
            Object {
              "fontSize": 14,
              "letterSpacing": 2,
              "textAlign": "right",
            },
            Object {
              "textAlign": "left",
            },
          ]
        }
        onPress={[Function]}
        style={
          Object {
            "marginHorizontal": 10,
            "marginTop": 5,
            "width": 80,
          }
        }
        text="CANCEL"
        type="transparent"
      />
    }
    right={
      <Button
        buttonTextStyle={
          Array [
            Object {
              "fontSize": 14,
              "letterSpacing": 2,
              "textAlign": "right",
            },
            Object {
              "textAlign": "right",
            },
          ]
        }
        onPress={[Function]}
        style={
          Object {
            "marginHorizontal": 10,
            "marginTop": 5,
            "width": 80,
          }
        }
        text="DONE"
        type="transparent"
      />
    }
    title="Status"
  />
  <Flex
    align="stretch"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
    value={1}
  >
    <TouchableIOS
      onPress={[Function]}
      pressProps={
        Array [
          "uncontacted",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#ECEEF2",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
    >
      <MyText
        style={
          Array [
            Object {
              "flex": 1,
              "fontSize": 16,
            },
            Object {
              "color": "#52C5DC",
            },
          ]
        }
      >
        Uncontacted
      </MyText>
      <Icon
        name="checkIcon"
        size={16}
        style={
          Object {
            "color": "#52C5DC",
          }
        }
        type="MissionHub"
      />
    </TouchableIOS>
    <TouchableIOS
      onPress={[Function]}
      pressProps={
        Array [
          "attempted_contact",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#ECEEF2",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
    >
      <MyText
        style={
          Array [
            Object {
              "flex": 1,
              "fontSize": 16,
            },
            null,
          ]
        }
      >
        Attempted Contact
      </MyText>
    </TouchableIOS>
    <TouchableIOS
      onPress={[Function]}
      pressProps={
        Array [
          "contacted",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#ECEEF2",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
    >
      <MyText
        style={
          Array [
            Object {
              "flex": 1,
              "fontSize": 16,
            },
            null,
          ]
        }
      >
        Contacted
      </MyText>
    </TouchableIOS>
    <TouchableIOS
      onPress={[Function]}
      pressProps={
        Array [
          "completed",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#ECEEF2",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
    >
      <MyText
        style={
          Array [
            Object {
              "flex": 1,
              "fontSize": 16,
            },
            null,
          ]
        }
      >
        Completed
      </MyText>
    </TouchableIOS>
    <TouchableIOS
      onPress={[Function]}
      pressProps={
        Array [
          "do_not_contact",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "#ECEEF2",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "paddingHorizontal": 20,
          "paddingVertical": 15,
        }
      }
    >
      <MyText
        style={
          Array [
            Object {
              "flex": 1,
              "fontSize": 16,
            },
            null,
          ]
        }
      >
        Do Not Contact
      </MyText>
    </TouchableIOS>
  </Flex>
</View>
`;
